src/tql/query.py

```diff
@@ -0,0 +1,185 @@
+    elif answer_type is NumericGroup:
+        return {group_key(numeric_group.owner()): decode(numeric_group.numeric(), Numeric) for numeric_group in answer}
+    else:
+        raise ValueError("Unknown answer type. Please report this error.")
+
+
+def run(connection, query, args, strict_transactions, global_inference):
+    query = query.strip()
+
+    if len(query) == 0:
+        return
+
+    query_args = "".join(query.split("\"")[::2]).replace(";", "").split()
+
+    keyword_counts = {
+        "match": 0,
+        "get": 0,
+        "define": 0,
+        "undefine": 0,
+        "insert": 0,
+        "delete": 0,
+        "group": 0,
+        "count": 0,
+        "sum": 0,
+        "max": 0,
+        "min": 0,
+        "mean": 0,
+        "median": 0,
+        "std": 0,
+    }
+
+    for arg in query_args:
+        if arg in keyword_counts:
+            keyword_counts[arg] += 1
+
+    aggregate_count = sum((
+        keyword_counts["count"],
+        keyword_counts["sum"],
+        keyword_counts["max"],
+        keyword_counts["min"],
+        keyword_counts["mean"],
+        keyword_counts["median"],
+        keyword_counts["std"],
+    ))
+
+    candidate_query_types = list()
+
+    if keyword_counts["group"] > 0 and aggregate_count > 0:
+        candidate_query_types.append("match_group_aggregate")
+    elif aggregate_count > 0:
+        candidate_query_types.append("match_aggregate")
+    elif keyword_counts["group"] > 0:
+        candidate_query_types.append("match_group")
+    elif keyword_counts["get"] > 0:
+        candidate_query_types.append("match")
+
+    if keyword_counts["define"] > 0:
+        candidate_query_types.append("define")
+
+    if keyword_counts["undefine"] > 0:
+        candidate_query_types.append("undefine")
+
+    if keyword_counts["insert"] > 0 and keyword_counts["delete"] > 0:
+        candidate_query_types.append("update")
+    elif keyword_counts["insert"] > 0:
+        candidate_query_types.append("insert")
+    elif keyword_counts["delete"] > 0:
+        candidate_query_types.append("delete")
+
+    if len(candidate_query_types) > 1:
+        raise QueryParsingError("Query contains incompatible keywords: '{}'".format("', '".join(candidate_query_types)))
+    elif len(candidate_query_types) == 1:
+        query_type = candidate_query_types[0]
+    elif keyword_counts["match"] > 0:
+        query_type = "match"
+    else:
+        raise QueryParsingError("Query contains no keywords.")
+
+    if args.session is None:
+        if strict_transactions:
+            raise ArgumentError("Strict transaction types is enabled and no session type was provided. Use -s to specify session type.")
+        elif query_type in ("define", "undefine"):
+            session_type = SessionType.SCHEMA
+        else:
+            session_type = SessionType.DATA
+    else:
+        if args.session.lower() == "schema":
+            session_type = SessionType.SCHEMA
+        elif args.session.lower() == "data":
```

Comment:
If there's any way for us to coerce these to lowers as we take them in as arguments rather than whenever we want to refer to them, that would be more ideal and easier to develop with in the future.

Reply:
We could, but for the `args` in general I don't think they should be modified in place from user inputs, as that has the potential to create generate errors without a clear cause. I think it's safer and clearer to keep the code for processing args together at the point of use, without in-place modification.

Reply:
I think we spoke about this further in person, do you remember if we made a decision?

Reply:
I don't think so. It doesn't seem like the scope of this project will get to the point where this matters, but it's something that's potentially easy to miss for anyone working on this in the future. If we're ever calling `.lower()` on the same arguments (session or transaction) after this point, that can be the appropriate time to make this change.

Reply:
Sounds good.